DynamicTablesPkg/Library/Acpi/Arm/AcpiTpm2LibArm/Tpm2Generator.c

```diff
@@ -0,0 +1,344 @@
+*/
+GET_OBJECT_LIST (
+  EObjNameSpaceArm,
+  EArmObjTpm2InterfaceInfo,
+  CM_ARM_TPM2_INTERFACE_INFO
+  )
+
+/** Construct the TPM2 ACPI table.
+
+  Called by the Dynamic Table Manager, this function invokes the
+  Configuration Manager protocol interface to get the required hardware
+  information for generating the ACPI table.
+
+  If this function allocates any resources then they must be freed
+  in the FreeTpm2TableResources function.
+
+  @param [in]  This           Pointer to the table generator.
+  @param [in]  AcpiTableInfo  Pointer to the ACPI Table Info.
+  @param [in]  CfgMgrProtocol Pointer to the Configuration Manager
+                              Protocol Interface.
+  @param [out] Table          Pointer to the constructed ACPI Table.
+
+  @retval EFI_SUCCESS           Table generated successfully.
+  @retval EFI_INVALID_PARAMETER A parameter is invalid.
+  @retval EFI_NOT_FOUND         The required object was not found.
+  @retval EFI_BAD_BUFFER_SIZE   The size returned by the Configuration
+                                Manager is less than the Object size for the
+                                requested object.
+  @retval EFI_OUT_OF_RESOURCES  Memory allocation failed.
+**/
+STATIC
+EFI_STATUS
+EFIAPI
+BuildTpm2Table (
+  IN  CONST ACPI_TABLE_GENERATOR                  *CONST  This,
+  IN  CONST CM_STD_OBJ_ACPI_TABLE_INFO            *CONST  AcpiTableInfo,
+  IN  CONST EDKII_CONFIGURATION_MANAGER_PROTOCOL  *CONST  CfgMgrProtocol,
+  OUT       EFI_ACPI_DESCRIPTION_HEADER          **CONST  Table
+  )
+{
+  EFI_STATUS                  Status;
+  UINT32                      TableSize;
+  CM_ARM_TPM2_INTERFACE_INFO  *TpmInfo;
+  EFI_TPM2_ACPI_TABLE_V4      *Tpm2;
+
+  *Table = NULL;
+
+  ASSERT (
+    (This != NULL) &&
+    (AcpiTableInfo != NULL) &&
+    (CfgMgrProtocol != NULL) &&
+    (Table != NULL) &&
+    (AcpiTableInfo->TableGeneratorId == This->GeneratorID) &&
+    (AcpiTableInfo->AcpiTableSignature == This->AcpiTableSignature)
+    );
+
+  if ((AcpiTableInfo->AcpiTableRevision < This->MinAcpiTableRevision) ||
+      (AcpiTableInfo->AcpiTableRevision > This->AcpiTableRevision))
+  {
+    DEBUG ((
+      DEBUG_ERROR,
+      "ERROR: TPM2: Requested table revision = %d is not supported. "
+      "Supported table revisions: Minimum = %d. Maximum = %d\n",
+      AcpiTableInfo->AcpiTableRevision,
+      This->MinAcpiTableRevision,
+      This->AcpiTableRevision
+      ));
+    return EFI_INVALID_PARAMETER;
+  }
+
+  Status = GetEArmObjTpm2InterfaceInfo (
+             CfgMgrProtocol,
+             CM_NULL_TOKEN,
+             &TpmInfo,
+             NULL
+             );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((
+      DEBUG_ERROR,
+      "%a: Failed to get TPM interface CM Object %r\n",
+      __func__,
+      Status
+      ));
+    return Status;
+  }
+
+  // Calculate the size of the TPM2 table
+  switch (AcpiTableInfo->AcpiTableRevision) {
+    case EFI_TPM2_ACPI_TABLE_REVISION_3:

```

Comment:
I didn't find the spec defining the revision 3 of the table,
would you have a link by any chance ?

Reply:
I thought I saw a reference to rev 3 somewhere but I can't find it anymore. I will change it to "if (AcpiTableInfo->AcpiTableRevision < EFI_TPM2_ACPI_TABLE_REVISION_4)" instead, following what had been done in Tcg2Acpi.c:
https://github.com/tianocore/edk2/blob/3b2025969e6e8a2f6542996182cd4132868641c6/SecurityPkg/Tcg/Tcg2Acpi/Tcg2Acpi.c#L824.

Reply:
The issue is that without the spec, it's hard to check the table generation is valid.
If we're not able to find the spec., it might be better to limit the generation of the table to the Revision 4 onward (and don't support the version 3).

Reply:
Makes sense. I will remove version 3 support then.